### PR TITLE
refactor(api): `opentrons.protocols.implementations` refactor

### LINKED ISSUE #7380: Refactor opentrons.protocols.implementations package
# Overview 

This is a complex package that needs reorganization, renaming, and other refactors. It was inspired by my hastily put [PR](https://github.com/Opentrons/opentrons/pull/7374).

### Naming

From slack thread @SyntaxColoring

First, throughout the opentrons/api source, I do feel a recurring “naming explosion” thing. For example, all of these packages are different:
`opentrons.protocols.implementations.interfaces`
`opentrons.protocols.implementations.protocol_api`
`opentrons.protocol_api`

I found (and still kind of find) this pretty overwhelming when I started. It’s hard for me to intuit what the gist of each of these packages is—and, more importantly, what the differences between them are, or should be.

##### Context
`implementations`, `implementations.interfaces`, `implementations.simulations`??? It isn't clear.

- `implementations` holds the classes that implement the protocol api. These are the guts of what used to be in the classes in `opentrons.protocol_api`. 
- `implementations.interfaces` are abstract base classes that are implemented in implementations`. Although `Well`, `WellGrid`, `PairedContext`, and `TipTracker` do not inherit from such interfaces.
- `implementations.simulations` holds implementations of `implementations.interfaces` classes for the purpose of simulation.

### Implementation leaking out

@mcous wrote:

Also, for this particular PR / area of the code, does it feel a little odd that the fact that a ProtocolContext is constructed to defer to an implementation class - which feels like an implementation detail to me - leaks out of the module?
To spitball in the absence of actually working through this from the top down, I wonder if something flatter that hides the implementation details could work:

```
from opentrons.protocol_api import create_protocol_context

context = create_protocol_context(
    # ...args and such
    # factory method decides what implementation to use based on args and FF's
)
```

##### Context

There are presently only two types of `ProtocolContextInterface` derived classes: `implementations.protocol_context.ProtocolContextSimulation` and `implementations.protocol_context.simulations.ProtocolContextSimulation`.

We will be adding another implementation that defers to the `ProtocolEngine`.

These are explicitly instantiated when creating `ProtocolContext` objects in our system. As @mcous mentioned this is an implementation detail leaking out of the package for no great reason.

### PR DESCRIPTION
**Pull request by amitlissack**
# Overview

Refactor of `opentrons.protocols.implementations` package. 

#### `opentrons.protocols.context`:
This is the root and public facing interface of context class implementations.
- abstract base classes 
- factory methods to create implementations

#### `opentrons.protocols.context.protocol_api`:
These are the implementations of the v2 protocol api

#### `opentrons.protocols.context.engine`:
These are the implementations that talk to protocol engine

#### `opentrons.protocols.context.simulator`:
This is the home of the simulating contexts, notably the `AbstractInstrument` implementation that enables "fast-sim".

closes #7380 

# Changelog

- renamed `implementations` package to `context`
- moved abstract base classes from `interfaces` to `context`, the root of the package.
- use `Abstract` instead of `Interface` in naming abstract base classes.
- Move `TipTracker` and `WellGrid` to `geometry` package.
- created `implementation_factory` module defining functions to create `AbstractLabware` and `AbstractProtocol` objects.

# Review requests

What do we think of the naming and structure?

# Risk assessment

Low. Mostly moving files around and renaming. The factory methods are covered by many tests.

### CODE REVIEW

**Review by sfoster1 — APPROVED** [review]
I like it!

File: api/src/opentrons/protocols/context/implementation_factory.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+from opentrons.types import Location
+from opentrons_shared_data.labware.dev_types import LabwareDefinition
+
+if TYPE_CHECKING:
+    from .protocol_api.protocol_context import ProtocolContextImplementation
+    from opentrons.protocol_engine.clients import SyncClient
+    from . import AbstractLabware, AbstractProtocol
+
+
+def create_labware(
```
Comment (mcous) [inline api/src/opentrons/protocols/context/implementation_factory.py]: What purpose is this factory serving? It doesn't appear to be hiding any implementation details nor doing anything with its arguments other than passing them verbatim to the constructor

File: api/src/opentrons/protocols/context/implementation_factory.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+from opentrons import APIVersion
+from opentrons.protocols.api_support.util import HardwareToManage
+from opentrons.protocols.types import Protocol
+
+from opentrons.types import Location
+from opentrons_shared_data.labware.dev_types import LabwareDefinition
+
+if TYPE_CHECKING:
+    from .protocol_api.protocol_context import ProtocolContextImplementation
```
Comment (mcous) [inline api/src/opentrons/protocols/context/implementation_factory.py]: From reading and these circular (type) dependencies, I'm wondering if the entire `opentrons.protocols.context` package would make more sense as part of `opentrons.protocol_api`, instead?

File: api/src/opentrons/protocols/context/implementation_factory.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        api_version=api_version,
+        hardware=hardware,
+        bundled_labware=bundled_labware,
+        bundled_data=bundled_data,
+        extra_labware=extra_labware,
+        protocol=protocol,
+    )
+
+
+def create_engine(client: SyncClient) -> AbstractProtocol:
```
Comment (mcous) [inline api/src/opentrons/protocols/context/implementation_factory.py]: Not sure what purpose this factory is serving, either. Could we maybe hold off until we have stuff to wire up?

File: api/src/opentrons/simulate.py [hunk 4]
```diff
@@ -195,7 +194,7 @@ def _build_protocol_context(
     version specification for use with
     :py:meth:`.protocol_api.execute.run_protocol`
     """
-    ctx_impl = ProtocolContextImplementation(
+    ctx_impl = implementation_factory.create_protocol(
```
Comment (mcous) [inline api/src/opentrons/simulate.py]: This feels to me like maybe the factory functions are one level too deep. The implementations are... implementation details.

I think it would feel more straightforward to me if this caller was using a factory to create a `ProtocolContext`, instead. That factory would do the work of spinning up the correct implementation based on args and settings instead of pushing that work to the caller

File: api/src/opentrons/protocols/context/implementation_factory.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
+"""Factory functions to create context implementation classes."""
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocols/context/implementation_factory.py]:
```suggestion
"""Factory functions to create concrete contexts."""
```

Would this be clearer?

**Review by mcous — COMMENTED** [review]
TL;DR:

- I like the switch to `Abstract` names
- All this stuff feels like "Protocol API" concerns rather than general protocol concerns, and given what's importing what, I wonder if we could put the implementations in `opentrons.protocol_api` instead?
- I think we might be able to get more value from the factory functions if we target the public protocol context class, instead
    - Since whatever factory functions we need are going to serve as integration points in the code, I think they're also difficult to get "right" from the bottom up
    - I'm down to remove or punt on this for the PR, but it is a discussion I want to have

---

I think our use of factory functions/methods should serve these main purposes:

- Wire up classes with their dependencies to hide implementation details and allow the classes themselves to remain loosely coupled
- Return different specific class instances of a parent class depending on input
    - Also probably to hide implementation details 

I think we should avoid factories that:

- Simply invoke a class constructor
    - Mercifully, Python does not use the `new` keyword
    - Avoiding coupling to `new` is the only reason I'd have a factory do this in the first place
- Only perform argument validation or parsing
    - If a factory is needed for argument parsing or validation that isn't proving any sort of implementation detail hiding, I'd argue that said validation belongs in the constructor, instead

---

If the purpose of implementing factory functions is to hide implementation details, then I think a few things fall out for me with this PR:

### what should we have factories for?

If we provide factory functions for the private implementation classes, we haven't really hidden any implementation details. **I think our factory (or factories) should return a full `ProtocolContext`**. Every usage of `create_protocol` and `create_protocol_simulation` I can find is immediately followed by the construction of a `ProtocolContext`, which to me is a sign that we went one layer too deep, because the implementation 

That way the factory can spin up whatever implementation is most appropriate given the arguments and feature flags present. I'm on the fence about whether we should have a separate factory for a simulating ProtocolContext, so I could be swayed either way.

### testing

In terms of testing, I think they should be primarily tested by integration or end-to-end tests

A factory's job is to wire up different systems and hide implementation details. When we talk about testing in general, I think tests drive value one of two main ways.

1. By checking "does this code work?" 
2. By providing architectural feedback by mapping out how different units interact

(1) would be accomplished by opaque testing of inputs and outputs of the subject. (2) would be accomplished by having the subject exercise test doubles. (2) also means that the tests are coupled to the implementation of the code, which increases friction to changing that code, because tests will fail if you change the implementation.

In this PR, the factories are tested by partially mocking out the modules from where the classes the factories will create are defined. So, we're not testing (1), because we're testing that the subject calls stuff the way we think it will, not that it's working at the end of the day.

But the trouble for me here is that I don't think (2) is really in play either. We've already got our structure in how we've split apart our classes and implementations; the factory is just here to wire it up properly. So what architectural value are we deriving from these tests?

My takeaway here is that if a factory function is in place to hide implementation details by wiring up dependencies, then the most valuable way to test it is via an integration or end-to-end test. I'd personally rather skip the test than mock out the implementation details we're supposed to be spinning up just to get the line coverage.

**Review by SyntaxColoring — APPROVED** [review]
Nice!

> - renamed `implementations` package to `context`

Would `context_backends` or `backends` be clearer?

- moved abstract base classes from `interfaces` to `context`, the root of the package.

I love it.

- use `Abstract` instead of `Interface` in naming abstract base classes.

Once I get over my Java trauma, I think I like it. "Interface" was a loaded term.

- Move `TipTracker` and `WellGrid` to `geometry` package.

I know nothing about that code, but it sounds surprising that a `TipTracker` would be in `geometry`.
 
- created `implementation_factory` module defining functions to create `AbstractLabware` and `AbstractProtocol` objects.

I vaguely agree with @mcous's opinions on factory functions in general, but I don't yet have an informed opinion on what those opinions mean for this PR. 🤔

**Comment by mcous** [comment]
Oh yeah, I like backends or context_backends, too 🚲

"Implementation" has the misfortune of meaning the right thing but being way to long to actually use in a module path

**Comment by SyntaxColoring** [comment]
> Oh yeah, I like backends or context_backends, too 🚲
> 
> "Implementation" has the misfortune of meaning the right thing but being way to long to actually use in a module path

My additional gripe with "implementation" is that it leads to apparent oxymorons like `implementations.interfaces.Foo` or `implementations.AbstractFoo`. (I thought, "'implementations' are the opposites of interfaces and abstract classes, aren't they?")

**Comment by amitlissack** [comment]
Cool. I'll get rid of the factory functions.
